File: chap17/VB/chap17/Edit.vb
Imports Autodesk.AutoCAD.ApplicationServices
Imports Autodesk.AutoCAD.DatabaseServices
Imports Autodesk.AutoCAD.EditorInput
Imports Autodesk.AutoCAD.Geometry
Imports Autodesk.AutoCAD.Runtime

Public Class Edit
    ' 移动的过程.
    Public Shared Sub Move(ByVal ent As Entity, ByVal sourcePt As Point3d, ByVal targetPt As Point3d)
        ' 移动的方向矢量.
        Dim vec As Vector3d = targetPt - sourcePt
        ' 移动的矩阵.
        Dim mt As Matrix3d = Matrix3d.Displacement(vec)
        ' 实施移动.
        ent.TransformBy(mt)
    End Sub

    ' 移动的过程.
    Public Shared Sub Move(ByVal id As ObjectId, ByVal sourcePt As Point3d, ByVal targetPt As Point3d)
        Dim mt As Matrix3d = Matrix3d.Displacement(targetPt - sourcePt)
        Dim db As Database = HostApplicationServices.WorkingDatabase
        Using trans As Transaction = db.TransactionManager.StartTransaction()
            Dim ent As Entity = trans.GetObject(id, OpenMode.ForWrite)
            ent.TransformBy(mt)
            trans.Commit()
        End Using
    End Sub

    ' 复制的过程.
    Public Shared Sub Copy(ByVal ent As Entity, ByVal sourcePt As Point3d, ByVal targetPt As Point3d)
        Dim mt As Matrix3d = Matrix3d.Displacement(targetPt - sourcePt)
        Dim entCopy As Entity = ent.GetTransformedCopy(mt)
        AppendEntity(entCopy)
    End Sub

    ' 复制的过程.
    Public Shared Sub Copy(ByVal id As ObjectId, ByVal sourcePt As Point3d, ByVal targetPt As Point3d)
        Dim mt As Matrix3d = Matrix3d.Displacement(targetPt - sourcePt)
        Dim db As Database = HostApplicationServices.WorkingDatabase
        Dim entCopy As Entity
        Using trans As Transaction = db.TransactionManager.StartTransaction()
            Dim ent As Entity = trans.GetObject(id, OpenMode.ForWrite)
            entCopy = ent.GetTransformedCopy(mt)
            trans.Commit()
        End Using
        AppendEntity(entCopy)
    End Sub

    ' 旋转的过程.
    Public Shared Sub Rotate(ByVal ent As Entity, ByVal basePt As Point3d, ByVal angle As Double)
        Dim mt As Matrix3d = Matrix3d.Rotation(angle, Vector3d.ZAxis, basePt)
        ent.TransformBy(mt)
    End Sub

    ' 旋转的过程.
    Public Shared Sub Rotate(ByVal id As ObjectId, ByVal basePt As Point3d, ByVal angle As Double)
        Dim mt As Matrix3d = Matrix3d.Rotation(angle, Vector3d.ZAxis, basePt)
        Dim db As Database = HostApplicationServices.WorkingDatabase
        Using trans As Transaction = db.TransactionManager.StartTransaction()
            Dim ent As Entity = trans.GetObject(id, OpenMode.ForWrite)
            ent.TransformBy(mt)
            trans.Commit()
        End Using
    End Sub

    ' 缩放的过程.
    Public Shared Sub Scale(ByVal ent As Entity, ByVal basePt As Point3d, ByVal scaleFactor As Double)
        Dim mt As Matrix3d = Matrix3d.Scaling(scaleFactor, basePt)
        ent.TransformBy(mt)
    End Sub

    ' 缩放的过程.
    Public Shared Sub Scale(ByVal id As ObjectId, ByVal basePt As Point3d, ByVal scaleFactor As Double)
        Dim mt As Matrix3d = Matrix3d.Scaling(scaleFactor, basePt)
        Dim db As Database = HostApplicationServices.WorkingDatabase
        Using trans As Transaction = db.TransactionManager.StartTransaction()
            Dim ent As Entity = trans.GetObject(id, OpenMode.ForWrite)
            ent.TransformBy(mt)
            trans.Commit()
        End Using
    End Sub

    ' 镜像的过程.
    Public Shared Sub Mirror(ByVal ent As Entity, ByVal mirrorPt1 As Point3d, ByVal mirrorPt2 As Point3d, ByVal eraseSourceObject As Boolean)
        ' 镜像线.
        Dim mirrorLine As New Line3d(mirrorPt1, mirrorPt2)
        ' 镜像矩阵.
        Dim mt As Matrix3d = Matrix3d.Mirroring(mirrorLine)
        If eraseSourceObject = True Then
            ent.TransformBy(mt)
        Else
            Dim entCopy As Entity = ent.GetTransformedCopy(mt)
            AppendEntity(entCopy)
        End If
    End Sub

    ' 镜像的过程.
    Public Shared Sub Mirror(ByVal id As ObjectId, ByVal mirrorPt1 As Point3d, ByVal mirrorPt2 As Point3d, ByVal eraseSourceObject As Boolean)
        Dim miLine As New Line3d(mirrorPt1, mirrorPt2)
        Dim mt As Matrix3d = Matrix3d.Mirroring(miLine)
        Dim db As Database = HostApplicationServices.WorkingDatabase

        Dim ent As Entity
        Using trans As Transaction = db.TransactionManager.StartTransaction()
            ent = trans.GetObject(id, OpenMode.ForWrite)
        End Using

        If eraseSourceObject = True Then
            ent.TransformBy(mt)
        Else
            Dim entCopy As Entity = ent.GetTransformedCopy(mt)
            AppendEntity(entCopy)
        End If
    End Sub

    ' 偏移的过程.
    Public Shared Sub Offset(ByVal cur As Curve, ByVal dis As Double)
        Dim db As Database = HostApplicationServices.WorkingDatabase
        Dim ed As Editor = Application.DocumentManager.MdiActiveDocument.Editor
        Try
            Dim offsetCur As DBObjectCollection = cur.GetOffsetCurves(dis)
            '将偏移的对象加入到数据库
            For i As Integer = 0 To offsetCur.Count - 1
                AppendEntity(offsetCur(i))
            Next
        Catch
            ed.WriteMessage("无法偏移！")
        End Try
    End Sub

    ' 偏移的过程.
    Public Shared Sub Offset(ByVal id As ObjectId, ByVal dis As Double)
        Dim db As Database = HostApplicationServices.WorkingDatabase
        Dim ed As Editor = Application.DocumentManager.MdiActiveDocument.Editor

        Using trans As Transaction = db.TransactionManager.StartTransaction()
            Dim ent As Entity = trans.GetObject(id, OpenMode.ForWrite)
            If TypeOf (ent) Is Curve Then
                Dim cur As Curve = ent
                Try
                    Dim offsetCur As DBObjectCollection = cur.GetOffsetCurves(dis)
                    '将偏移的对象加入到数据库
                    For i As Integer = 0 To offsetCur.Count - 1
                        AppendEntity(offsetCur(i))
                    Next
                    trans.Commit()
                Catch
                    ed.WriteMessage("无法偏移！")
                End Try
            Else
                ed.WriteMessage("无法偏移！")
            End If
        End Using
    End Sub

    ' 矩形阵列的过程.
    Public Shared Sub ArrayRectang(ByVal ent As Entity, ByVal numRows As Integer, ByVal numCols As Integer, ByVal disRows As Double, ByVal disCols As Double)
        Dim db As Database = HostApplicationServices.WorkingDatabase
        Using trans As Transaction = db.TransactionManager.StartTransaction()
            For m As Integer = 0 To numRows - 1
                For n As Integer = 0 To numCols - 1
                    Dim mt As Matrix3d = Matrix3d.Displacement(New Vector3d(n * disCols, m * disRows, 0))
                    Dim entCopy As Entity = ent.GetTransformedCopy(mt)
                    AppendEntity(entCopy)
                Next
            Next
            ' 删除多余的对象.
            ent.Erase()
            trans.Commit()
        End Using
    End Sub

    ' 矩形阵列的过程.
    Public Shared Sub ArrayRectang(ByVal id As ObjectId, ByVal numRows As Integer, ByVal numCols As Integer, ByVal disRows As Double, ByVal disCols As Double)
        Dim db As Database = HostApplicationServices.WorkingDatabase
        Using trans As Transaction = db.TransactionManager.StartTransaction()
            Dim ent As Entity = trans.GetObject(id, OpenMode.ForWrite)
            For m As Integer = 0 To numRows - 1
                For n As Integer = 0 To numCols - 1
                    Dim mt As Matrix3d = Matrix3d.Displacement(New Vector3d(n * disCols, m * disRows, 0))
                    Dim entCopy As Entity = ent.GetTransformedCopy(mt)
                    AppendEntity(entCopy)
                Next
            Next
            ' 删除多余的对象.
            ent.Erase()
            trans.Commit()
        End Using
    End Sub

    ' 环形阵列的过程.
    Public Shared Sub ArrayPolar(ByVal ent As Entity, ByVal cenPt As Point3d, ByVal numObj As Integer, ByVal Angle As Double)
        Dim db As Database = HostApplicationServices.WorkingDatabase
        Using trans As Transaction = db.TransactionManager.StartTransaction()
            For i As Integer = 0 To numObj - 2
                Dim mt As Matrix3d = Matrix3d.Rotation(Angle * (i + 1) / numObj, Vector3d.ZAxis, cenPt)
                Dim entCopy As Entity = ent.GetTransformedCopy(mt)
                AppendEntity(entCopy)
            Next
            trans.Commit()
        End Using
    End Sub

    ' 环形阵列的过程.
    Public Shared Sub ArrayPolar(ByVal id As ObjectId, ByVal cenPt As Point3d, ByVal numObj As Integer, ByVal Angle As Double)
        Dim db As Database = HostApplicationServices.WorkingDatabase
        Using trans As Transaction = db.TransactionManager.StartTransaction()
            Dim ent As Entity = trans.GetObject(id, OpenMode.ForWrite)
            For i As Integer = 0 To numObj - 2
                Dim mt As Matrix3d = Matrix3d.Rotation(Angle * (i + 1) / numObj, Vector3d.ZAxis, cenPt)
                Dim entCopy As Entity = ent.GetTransformedCopy(mt)
                AppendEntity(entCopy)
            Next
            trans.Commit()
        End Using
    End Sub

    ' 删除的过程.
    Public Shared Sub [Erase](ByVal ent As Entity)
        ent.Erase()
    End Sub

    ' 删除的过程.
    Public Shared Sub [Erase](ByVal id As ObjectId)
        Dim db As Database = HostApplicationServices.WorkingDatabase
        Using trans As Transaction = db.TransactionManager.StartTransaction()
            Dim ent As Entity = trans.GetObject(id, OpenMode.ForWrite)
            ent.Erase()
            trans.Commit()
        End Using
    End Sub

    ' 度化弧度的函数.
    Public Shared Function Rad2Ang(ByVal angle As Double) As Double
        Dim pi As Double = 4 * Math.Atan(1)
        Rad2Ang = angle * pi / 180
    End Function

    ' 将图形对象加入到模型空间的函数.
    Public Shared Function AppendEntity(ByVal obj As Entity) As ObjectId
        Dim db As Database = HostApplicationServices.WorkingDatabase
        Dim objId As ObjectId
        Using trans As Transaction = db.TransactionManager.StartTransaction
            Dim bt As BlockTable = trans.GetObject(db.BlockTableId, OpenMode.ForRead)
            Dim btr As BlockTableRecord = trans.GetObject(bt(BlockTableRecord.ModelSpace), OpenMode.ForWrite)
            objId = btr.AppendEntity(obj)
            trans.AddNewlyCreatedDBObject(obj, True)
            trans.Commit()
        End Using
        Return objId
    End Function
End Class

File: chap17/C#/chap17/Edit.cs
using System;
using Autodesk.AutoCAD.ApplicationServices;
using Autodesk.AutoCAD.Colors;
using Autodesk.AutoCAD.DatabaseServices;
using Autodesk.AutoCAD.EditorInput;
using Autodesk.AutoCAD.Geometry;
using Autodesk.AutoCAD.Runtime;

namespace chap17
{
    public class Edit
    {
        // 移动的函数.
        public static void Move(Entity ent, Point3d sourcePt, Point3d targetPt)
        {
            Vector3d vec = targetPt - sourcePt;
            Matrix3d mt = Matrix3d.Displacement(vec);
            ent.TransformBy(mt);
        }

        // 移动的函数.
        public static void Move(ObjectId id, Point3d sourcePt, Point3d targetPt)
        {
            Matrix3d mt = Matrix3d.Displacement(targetPt - sourcePt);
            Database db = HostApplicationServices.WorkingDatabase;
            using (Transaction trans = db.TransactionManager.StartTransaction())
            {
                Entity ent = (Entity)trans.GetObject(id, OpenMode.ForWrite);
                ent.TransformBy(mt);
                trans.Commit();
            }
        }

        // 复制的函数.
        public static void Copy(Entity ent, Point3d sourcePt, Point3d targetPt)
        {
            Matrix3d mt = Matrix3d.Displacement(targetPt - sourcePt);
            Entity entCopy = ent.GetTransformedCopy(mt);
            AppendEntity(entCopy);
        }

        // 复制的函数.
        public static void Copy(ObjectId id, Point3d sourcePt, Point3d targetPt)
        {
            Matrix3d mt = Matrix3d.Displacement(targetPt - sourcePt);
            Database db = HostApplicationServices.WorkingDatabase;
            Entity entCopy;
            using (Transaction trans = db.TransactionManager.StartTransaction())
            {
                Entity ent = (Entity)trans.GetObject(id, OpenMode.ForWrite);
                entCopy = ent.GetTransformedCopy(mt);
                trans.Commit();
            }
            AppendEntity(entCopy);
        }

        // 旋转的函数.
        public static void Rotate(Entity ent, Point3d basePt, Double angle)
        {
            Matrix3d mt = Matrix3d.Rotation(angle, Vector3d.ZAxis, basePt);
            ent.TransformBy(mt);
        }

        // 旋转的函数.
        public static void Rotate(ObjectId id, Point3d basePt, Double angle)
        {
            Matrix3d mt = Matrix3d.Rotation(angle, Vector3d.ZAxis, basePt);
            Database db = HostApplicationServices.WorkingDatabase;
            using (Transaction trans = db.TransactionManager.StartTransaction())
            {
                Entity ent = (Entity)trans.GetObject(id, OpenMode.ForWrite);
                ent.TransformBy(mt);
                trans.Commit();
            }
        }

        // 缩放的函数.
        public static void Scale(Entity ent, Point3d basePt, Double scaleFactor)
        {
            Matrix3d mt = Matrix3d.Scaling(scaleFactor, basePt);
            ent.TransformBy(mt);
        }

        // 缩放的函数.
        public static void Scale(ObjectId id, Point3d basePt, Double scaleFactor)
        {
            Matrix3d mt = Matrix3d.Scaling(scaleFactor, basePt);
            Database db = HostApplicationServices.WorkingDatabase;
            Entity ent;
            using (Transaction trans = db.TransactionManager.StartTransaction())
            {
                ent = (Entity)trans.GetObject(id, OpenMode.ForWrite);
                ent.TransformBy(mt);
                trans.Commit();
            }
        }

        // 镜像的函数.
        public static void Mirror(Entity ent, Point3d mirrorPt1, Point3d mirrorPt2, bool eraseSourceObject)
        {
            Line3d mirrorLine = new Line3d(mirrorPt1, mirrorPt2);

            Matrix3d mt = Matrix3d.Mirroring(mirrorLine);
            if (eraseSourceObject == true)
                ent.TransformBy(mt);
            else
            {
                Entity entCopy = ent.GetTransformedCopy(mt);
                AppendEntity(entCopy);
            }
        }

        // 镜像的函数.
        public static void Mirror(ObjectId id, Point3d mirrorPt1, Point3d mirrorPt2, bool eraseSourceObject)
        {
            Line3d miLine = new Line3d(mirrorPt1, mirrorPt2);
            Matrix3d mt = Matrix3d.Mirroring(miLine);
            Database db = HostApplicationServices.WorkingDatabase;

            Entity ent;
            using (Transaction trans = db.TransactionManager.StartTransaction())
            {
                ent = (Entity)trans.GetObject(id, OpenMode.ForWrite);
                if (eraseSourceObject == true)
                    ent.TransformBy(mt);
                else
                {
                    Entity entCopy = ent.GetTransformedCopy(mt);
                    AppendEntity(entCopy);
                }
                trans.Commit();
            }
        }

        // 偏移的函数.
        public static void Offset(Curve cur, Double dis)
        {
            Database db = HostApplicationServices.WorkingDatabase;
            Editor ed = Application.DocumentManager.MdiActiveDocument.Editor;
            try
            {
                DBObjectCollection offsetCur = cur.GetOffsetCurves(dis);
                for (int i = 0; i < offsetCur.Count; i++)
                    AppendEntity((Entity)offsetCur[i]);
            }
            catch
            {
                ed.WriteMessage("无法偏移！");
            }
        }

        // 偏移的函数.
        public static void Offset(ObjectId id, Double dis)
        {
            Database db = HostApplicationServices.WorkingDatabase;
            Editor ed = Application.DocumentManager.MdiActiveDocument.Editor;

            using (Transaction trans = db.TransactionManager.StartTransaction())
            {
                Entity ent = (Entity)trans.GetObject(id, OpenMode.ForWrite);
                if (ent is Curve)
                {
                    Curve cur = (Curve)ent;
                    try
                    {
                        DBObjectCollection offsetCur = cur.GetOffsetCurves(dis);
                        for (int i = 0; i < offsetCur.Count; i++)
                            AppendEntity((Entity)offsetCur[i]);
                        trans.Commit();
                    }
                    catch
                    {
                        ed.WriteMessage("\n无法偏移！");
                    }
                }
                else
                    ed.WriteMessage("\n无法偏移！");
            }
        }

        // 矩形阵列的函数.
        public static void ArrayRectang(Entity ent, int numRows, int numCols, double disRows, double disCols)
        {
            Database db = HostApplicationServices.WorkingDatabase;
            using (Transaction trans = db.TransactionManager.StartTransaction())
            {
                for (int m = 0; m < numRows; m++)
                {
                    for (int n = 0; n < numCols; n++)
                    {
                        Matrix3d mt = Matrix3d.Displacement(new Vector3d(n * disCols, m * disRows, 0));
                        Entity entCopy = ent.GetTransformedCopy(mt);
                        AppendEntity(entCopy);
                    }
                }
                ent.Erase();
                trans.Commit();
            }
        }

        // 矩形阵列的函数.
        public static void ArrayRectang(ObjectId id, int numRows, int numCols, double disRows, double disCols)
        {
            Database db = HostApplicationServices.WorkingDatabase;
            using (Transaction trans = db.TransactionManager.StartTransaction())
            {
                Entity ent = (Entity)trans.GetObject(id, OpenMode.ForWrite);
                for (int m = 0; m < numRows; m++)
                {
                    for (int n = 0; n < numCols; n++)
                    {
                        Matrix3d mt = Matrix3d.Displacement(new Vector3d(n * disCols, m * disRows, 0));
                        Entity entCopy = (Entity)ent.GetTransformedCopy(mt);
                        AppendEntity(entCopy);
                    }
                }
                ent.Erase();
                trans.Commit();
            }
        }

        // 环形阵列的函数.
        public static void ArrayPolar(Entity ent, Point3d cenPt, int numObj, double Angle)
        {
            Database db = HostApplicationServices.WorkingDatabase;
            using (Transaction trans = db.TransactionManager.StartTransaction())
            {
                for (int i = 0; i < numObj - 1; i++)
                {
                    Matrix3d mt = Matrix3d.Rotation(Angle * (i + 1) / numObj, Vector3d.ZAxis, cenPt);
                    Entity entCopy = ent.GetTransformedCopy(mt);
                    AppendEntity(entCopy);
                }
                trans.Commit();
            }
        }

        // 环形阵列的函数.
        public static void ArrayPolar(ObjectId id, Point3d cenPt, int numObj, double Angle)
        {
            Database db = HostApplicationServices.WorkingDatabase;
            using (Transaction trans = db.TransactionManager.StartTransaction())
            {
                Entity ent = (Entity)trans.GetObject(id, OpenMode.ForWrite);
                for (int i = 0; i < numObj - 1; i++)
                {
                    Matrix3d mt = Matrix3d.Rotation(Angle * (i + 1) / numObj, Vector3d.ZAxis, cenPt);
                    Entity entCopy = ent.GetTransformedCopy(mt);
                    AppendEntity(entCopy);
                }
                trans.Commit();
            }
        }

        // 删除的函数.
        public static void Erase(Entity ent)
        {
            ent.Erase();
        }

        // 删除的函数.
        public static void Erase(ObjectId id)
        {
            Database db = HostApplicationServices.WorkingDatabase;
            using (Transaction trans = db.TransactionManager.StartTransaction())
            {
                Entity ent = (Entity)trans.GetObject(id, OpenMode.ForWrite);
                ent.Erase();
                trans.Commit();
            }
        }

        // 度化弧度的函数.
        public static double Rad2Ang(double angle)
        {
            double rad = angle * Math.PI / 180;
            return rad;
        }

        // 将图形对象加入到模型空间的函数.
        public static ObjectId AppendEntity(Entity ent)
        {
            Database db = HostApplicationServices.WorkingDatabase;
            ObjectId entId;
            using (Transaction trans = db.TransactionManager.StartTransaction())
            {
                BlockTable bt = (BlockTable)trans.GetObject(db.BlockTableId, OpenMode.ForRead);
                BlockTableRecord btr = (BlockTableRecord)trans.GetObject(bt[BlockTableRecord.ModelSpace], OpenMode.ForWrite);
                entId = btr.AppendEntity(ent);
                trans.AddNewlyCreatedDBObject(ent, true);
                trans.Commit();
            }
            return entId;
        }
    }
}
